src/azstoragetorch/io.py

```diff
@@ -0,0 +1,188 @@
+# -------------------------------------------------------------------------
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT License. See LICENSE in the project root for
+# license information.
+# --------------------------------------------------------------------------
+
+import io
+import os
+from typing import get_args, Optional, Union, Literal
+import urllib.parse
+
+from azure.identity import DefaultAzureCredential
+from azure.core.credentials import (
+    AzureSasCredential,
+    TokenCredential,
+)
+
+
+_SUPPORTED_MODES = Literal["rb"]
+_SDK_CREDENTIAL_TYPE = Optional[
+    Union[
+        AzureSasCredential,
+        TokenCredential,
+    ]
+]
+_AZSTORAGETORCH_CREDENTIAL_TYPE = Union[_SDK_CREDENTIAL_TYPE, Literal[False]]
+
+
+class BlobIO(io.IOBase):
+    def __init__(
+        self,
+        blob_url: str,
+        mode: _SUPPORTED_MODES,
+        *,
+        credential: _AZSTORAGETORCH_CREDENTIAL_TYPE = None,
+    ):
+        self._blob_url = blob_url
+        self._validate_mode(mode)
+        self._mode = mode
+        self._sdk_credential = self._get_sdk_credential(blob_url, credential)
+
+        self._blob_downloader = self._get_blob_downloader()
+        self._position = 0
+        self._closed = False
+
+    def close(self) -> None:
+        self._close_blob_downloader()
+        self._closed = True
+
+    @property
+    def closed(self) -> bool:
+        return self._closed
+
+    def fileno(self) -> int:
+        raise OSError("BlobIO object has no fileno")
+
+    def flush(self) -> None:
+        self._validate_not_closed()
+
+    def read(self, size: Optional[int] = -1, /) -> bytes:
```

Comment:
Does size need to be `Optional` here? Would `None` be different than `-1`?

Or is my understanding of `Optional` not right. I always thought it just expands to `Union[<type>, None]` in which case `int` would be sufficient here.

Reply:
Yeah optional is still needed here. Technically, someone can pass in `None`:
```
blob_io.read(None)
```
and explicitly `-1`:
```
blob_io.read(-1)
```
If the `Optional` was not there, it would fail mypy checking if someone passed in `None` because `None` is not captured by `int` type. It also how typeshed types some of the `io` implementation as well. For example, this is the typing for BufferedIOBase: https://github.com/python/typeshed/blob/af4df4eb24f1cda42033f74440ba1117ce4c8f7e/stdlib/_io.pyi#L62